### PR TITLE
feat(webpack): add webpack to viper-news

### PR DESCRIPTION
**Pull request by TheLarkInn**
[WIP] Work in progress, still have a few pieces to clean up 😎 [WIP]

* This PR removes browserify, and replaces it with webpack. ~More info to appended to PR to come.~
* As part of this PR, the about page and the template and javascript associated is now async loaded

Load times _appear_ to be cut in half by these changes:

**Before**
screen shot 2017-06-27 at 2 51 33 pm

**After**
screen shot 2017-06-27 at 2 52 38 pm

I wanted to try and do the same for the comments and user section but couldn't spend much more time on it. :joy:

### CODE REVIEW

**Comment by WebReflection** [comment]
Hey @TheLarkInn, first of all, thanks a lot for the PR and the effort.

There are a couple of things though, I'd like to understand:

  * I don't want to transpile the `/sw.js`, just eventually minify it. That file is loaded **only** by ServiceWorker capable browsers. Babel has no place there, at least not for ES2015. Is it possible to do so?
  * the `sw.js` file should not be bundled. I am not sure I am missing the bit where it clearly gets not bundled
  * the whole point here is to demonstrate that both NodeJS and Browser can share (and eventually adopt on the browser) same exact template. However, the way I was bringing in the `bundle.js` before was in the `index.html` file but I don't see any change in the loader. Is that automagically brought in as soon as I just load in the `bundle.js` ? What I mean, is that on modern browsers, but also on non SW capable one, the whole client-side app should just work. Is that the case already?

Thanks a lot for any sort of clarification. I have to admit I've never looked into Webpack and if all my concerns/questions are answered and it's really that faster on load I guess no other library would ever possibly match the `viperHTML` + `hyperHTML` combo :tada: 

It's already the fastest in that page, FWIW, can't believe it can be even faster!

Thanks again.

**Comment by TheLarkInn** [comment]
> I don't want to transpile the /sw.js, just eventually minify it. That file is loaded only by ServiceWorker capable browsers. Babel has no place there, at least not for ES2015. Is it possible to do so? 

Yes I did notice in the diff the "use strict" was wiped maybe that was an error on my part. webpack should not grab the SW unless it's somewhere explicitly stated in the client entry code. I can fix the use strict omissions. Also it should not be in the bundle at all also. In the PR I set the --verbose flag which can give insight on to each module and why it is inside the bundle.

> the whole point here is to demonstrate that both NodeJS and Browser can share (and eventually adopt on the browser) same exact template. However, the way I was bringing in the bundle.js before was in the index.html file but I don't see any change in the loader. Is that automagically brought in as soon as I just load in the bundle.js ? What I mean, is that on modern browsers, but also on non SW capable one, the whole client-side app should just work. Is that the case already?

Yes! So I didn't mess with any of the isomorphic code although converting to esm was tempting it would mean I'd have to webpack the server too (which is more than doable with `target: node`.) Now if you are curious about the second bundle, `bundle.0.js` this is a "lazy chunk". If you notice the import() statement I've used in the renderer, webpack see's this and statically analyzes all sub deps of that module and places them in a separate bundle loaded lazily. webpack when this module is needed will jsonP the resource and append the proper script tag when it's time to load that code. Thus, lessening the initial js impact.

As fast as I can tell I didn't edit the index.html or loader so it should still serve the same. (Unless it was on accident 😂)

**Comment by TheLarkInn** [comment]
- [ ] remove renegade .DS_Store
- [ ] find out why use strict got stripped (babel?)

**Comment by WebReflection** [comment]
OK, running it locally looks like things are working as expected indeed.

However, if you run this version VS the previous one via Lighthouse, you'll see this version is slower.

Like ... 100ms slower on bootstrap.

Do you think there's anything we could improve to speed up the first paint?

Here numbers:

### Webpack on localhost:3000

Cold start, browser with just one tab

screenshot from 2017-06-28 09-19-13


### Browserify on localhost:3000

Cold start, browser with just one tab

screenshot from 2017-06-28 09-21-10


I know the 100ms difference might not be always that big but I've noticed browserify version was consistently around 700ms while this webpack is more around 800.

It's not a huge difference in the real-world out there but ... you know, if I need to rule out browserify and obtain worse results than before ... I'd not be too happy :stuck_out_tongue: 

Thanks for any sort of outcome or improvement, I'm also learning few things about Webpack here.

Cheers

**Comment by WebReflection** [comment]
P.S. I've also just realized when I've taken the Webpack screenshot it was on a second run and the difference went down to just 50ms slower, not 100ms like it was on first run.

I just wonder if there's really nothing else we could try to improve first paint perf. After all the SSR version should not even be impacted by these JS changes, right?

**Comment by WebReflection** [comment]
Moar benchmarks coming.

### Webpack network

screenshot from 2017-06-28 09-36-14


### Browserify network

screenshot from 2017-06-28 09-35-34

**Comment by WebReflection** [comment]
OK, hopefully last comment of mine for today.

Turns out, I actually need the body to be there, or at least the main section.

Please feel free to move that index.html script loader around or even remove it completely.

As you can see [in here](https://github.com/WebReflection/viper-news/commit/53d1d6b5b472eaafe9f94f341ea1cce797827098#diff-87cc76af36f1f4ad0ca0be68eaf8c309R101), not long time ago I was loading the SW after having the bundle in, and only [once DOMContentLoaded was triggered](https://github.com/WebReflection/viper-news/commit/53d1d6b5b472eaafe9f94f341ea1cce797827098#diff-9ac9992a17a643963f7a9b16f14ef6b6L16).

If you think Webpack can do a better job with its own way to load the bundle, I'd be happy to put back that `if` statement and have `hyperHTML` running later on, as long as the first paint is low :wink: 

Apologies for all the troubles :smile:

**Comment by TheLarkInn** [comment]
No troubles whatsoever. Im in USA (CST) so timezones obviously different. 

So essentially heres what I'm trying to accomplish: 
I noticed that in the original single bundle, all of the markup "wires?" information are available in initial bundle. My theory is that you could reduce your initial load time by "splitting out" the code for separate views into async "jsonp-fetched" bundles. 

webpack does this when you use `import()` in your JavaScript, and will split any sub dependencies of that module into a separate `chunk`, and converts that `import()` statement into a `Promise` that returns the module (as if you were `import` ing it). 

I tried applying this with the least amount of impact (not knowing how viper/hyper) handles async rendering for example: 

```
const asyncRender = (pagePromise) => (wire, model) => 
  pagePromise.then(renderFn => renderFn(wire, model))

const aboutPagePromise = import("../../shared/view/about");

const view = {
  about: (wire, model) => asyncRender(aboutPagePromise)(wire, model),
  summary: require('../../shared/view/summary'),
  item: require('../../shared/view/item'),
  comment: require('../../shared/view/comment'),
  user: require('../../shared/view/user'),
  next: require('../../shared/view/next')
};
```

Not knowing (intimately) viper/hyper, I'm not sure if there is support for this. An example might be how in VueJS, if I wanted to apply a component, all I have to do is pass a function that returns a Promise. (aka use `import()`);

Converting code that would look like:
```
import Component from "./my/component.vue"
Vue.component("some-component-tag", Vue.extend(Component))
```
   
into its Async counterpart.

```
const Component = () => import("./my/component.vue")

Vue.component("some-component-tag", Component")
```

Now if there is a way that this can be handled out of the box? (viperHTML.async() ??) The part that allows webpack to shine is having a imperative DSL that _is_ valid syntax at the same time to mark dependencies as async. So I want to try and leverage that if possible to really reduce the size of the bundle.

**Comment by TheLarkInn** [comment]
Oh and in regards to the lazy loading the SW: I can look at that also. Really want to find out the right way to handle the async pieces of (hy/vi)per first.

**Comment by TheLarkInn** [comment]
Oh and also: for [this comment](https://github.com/WebReflection/viper-news/pull/2#issuecomment-311594781), I notice that both bundles were being downloaded on initial load, were you testing this on the about page or main page. I wasn't seeing the second bundle even being requested on main page when I tested. I will attempt again to make sure.

**Comment by WebReflection** [comment]
the double download is due ServiceWorker, I am not sure there's anything we can do about it 'cause AFAIK it uses a different network stack.

#### Off Topic
I'm experimenting with the most basic "_shell_" to develop `viperHTML` based apps and it's based on Webpack 'cause I really liked it:
https://github.com/WebReflection/create-viperhtml-app

#### Back On Topic
As you can see in [this example file](https://github.com/WebReflection/create-viperhtml-app/blob/master/viper/server/index.js#L26-L48) from that experimental repo, `viperHTML` has zero issues in piping the layout through asynchronous code / promises.

However, I've never thought about implementing the same on `hyperHTML` side, for the simple reason it's so fast you can simply invoke a render again as soon as any promise has been resolved.

Moreover, `hyperHTML.bind(element)` and `hyperHTML.wire(object)` returns DOM nodes but I could create another kind of wire that is specifically `async` and return a placeholder to update once the Promise is resolved.

That being said, having two different meaning for `.async` on client and server doesn't seem to be ideal, so I'll think about it and see what I can do.

TL;DR I see the benefits of accepting asynchronous chunks on the client-side too, let me work on that.

Thanks

**Comment by WebReflection** [comment]
Right ... I'm wrapping a bit my head around this.

If I understand correctly, what is needed is not an asynchronous content, like it's possible already in `viperHTML`, but an **asynchronous render** (or even both one day)

am I correct?

**Comment by WebReflection** [comment]
FYI: I've just published `v0.15` of hyperHTML, here you are on `0.14`.

The main difference is that html content, represented by those interpolations that don't have surrounding spaces or other chars like it is already for `about`, `summary`, `item`, `comment`, `user`, and `next` parts of the view (in the render.js file), can now all load once on demand the same view and once resolved, return their sub-rendered content.

I hope this works!

**Comment by TheLarkInn** [comment]
This is awesome!!!!! Sorry I've been async from work. So what would the interface look like for where I would pass a ()=> promise with resolved "bound" content

**Comment by WebReflection** [comment]
Current status with `hyperHTML` version `0.15.1` and this change in the render.js

```js
const asyncRender = (pagePromise, wire, model) =>
  pagePromise.then(renderFn => renderFn(wire, model));

const aboutPagePromise = import("../../shared/view/about");
const summaryPagePromise = import("../../shared/view/summary");
const itemPagePromise = import("../../shared/view/item");
const commentPagePromise = import("../../shared/view/comment");
const userPagePromise = import("../../shared/view/user");
const nextPagePromise = import("../../shared/view/next");

const view = {
  about: (wire, model) => asyncRender(aboutPagePromise, wire, model),
  summary: (wire, model) => asyncRender(summaryPagePromise, wire, model),
  item: (wire, model) => asyncRender(itemPagePromise, wire, model),
  comment: (wire, model) => asyncRender(commentPagePromise, wire, model),
  user: (wire, model) => asyncRender(userPagePromise, wire, model),
  next: (wire, model) => asyncRender(nextPagePromise, wire, model)
};
```

Current result on the network panel:

screenshot from 2017-06-28 18-26-55

Looks like it was easier than I thought, and definitively successful.

You just pass along wires as promises and you render whenever it's possible :tada:

**Comment by WebReflection** [comment]
Writing down a couple of things before merging, deploying, publishing results, etc ...

  * considering I've resolved the template bit, is there anything else we can do to speed up boot time?
  * should I tell the ServiceWorker to wait and cache all bundles or the first one should be enough?
  * in case all bundles are needed, should I automate such procedure?
  * why are all bundles coming through anyway? is that expected?

If I can move on from here by my own I wouldn't bother you any extra time, you've done already a lot!!!

Please let me know if there's anything in particular you'd like to test/push/play with, thanks.

**Comment by TheLarkInn** [comment]
Haha that's so fast!!!! Yes so ideally you precache any "route" or view aka lazy view that isn't immediately needed for the page. With mobile phones and depending on your server and if you support h2 or not you might change or tweak the code splitting. Seeing as this is using H2, I'd imagine having smaller pieces is very beneficial. But if it was http/1 I'd go less granular.

**Comment by TheLarkInn** [comment]
Let me get the PR updated!!!!!!!!!!!!

**Comment by WebReflection** [comment]
..... fffffine ... I'll wait ... :laughing:

**Comment by TheLarkInn** [comment]
Now that I've verified with webpack-bundle-analzyer, this is looking really awesome. 
screen shot 2017-06-28 at 1 10 28 pm

**Comment by TheLarkInn** [comment]
Alrightyyy!!! Here we go. I updated the SW to precache the existing routes as well. We could add link rel="preload" if we wanted to but it gets hairier with duplicate network calls, etc. In this case this should be a great start. 

Since my computer is active funny, can you verify what kind of lighthouse and performance wins you are seeing on your local http/2 env?

File: client/js/render.js [hunk 1]
```diff
@@ -1,12 +1,15 @@
 const hyperHTML = require('hyperhtml');
 
+const asyncRenderView = (viewName, wire, model) =>
+  import(`../../shared/view/${viewName}.js`).then(renderFn => renderFn(wire, model));
```
Comment (TheLarkInn) [inline client/js/render.js:4]: So this is a clever feature of webpack called ContextModules.

We basically can treat a "context" (aka directory), as a module. What this allows is for webpack to resolve every module that matches that pattern in that partial path, and create separate async bundles out of each of those modules. This supports "pseudo" dynamic (via static analysis) loading.

**Comment by TheLarkInn** [comment]
I have a tiny feature we should add but needed to test it so it could come after you submit to HNPWA.com

**Comment by WebReflection** [comment]
I'm publishing and also changing locally the lazy loader ... I'm moving the SW check inside the bundle

**Comment by WebReflection** [comment]
FWI: current merge is live here:
https://viperhtml-164315.appspot.com/top/1

**Comment by TheLarkInn** [comment]
Nice! That's what I was just trying myself too. I was curious what was causing the network delay for the main script.

**Comment by TheLarkInn** [comment]
Ahhhhhhh we should preload the bundle I was seeing a huge win by dropping bundle down from that loader into bottom of body, and then also adding the following in the head 


```
    <link rel="preload" href="/js/bundle.js" as="script"> 

```

**Comment by WebReflection** [comment]
Current lighthouse results in here:
blob:chrome-extension://blipmdconlkpinefehnmjammfjpmpbjk/3857af1b-7faa-4976-ac6e-718e9c476c3c

I'm deploying the version with the following layout instead:
```html
<!doctype html>
<html lang="en">
  <head>
    <title> ${model.title} </title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="theme-color" content="#ffffff">
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="/favicon-16x16.png" sizes="16x16">
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
    <link rel="manifest" href="/manifest.json">
    <style>${model.style}</style>
  </head>
  <body>
    <header>${model.header}</header>
    <main>${model.main}</main>
    <script async src="/js/bundle.js"></script>
  </body>
</html>
```

**Comment by TheLarkInn** [comment]
Beautiful!!! Would you add `<link rel="preload" href="/js/bundle.js" as="script">` also for that resource. I believe it shouldn't be render blocking. (unless I'm mistaken)

**Comment by TheLarkInn** [comment]
I'll test it after you commit, ship and deploy and add another pr if needed.

**Comment by WebReflection** [comment]
I'll try, let's see this one without it how it goes.
My concern is that no mater when we load the first script, viperHTML renders the whole app exactly as it is even without JS.

Maybe there's something else I could do on the CSS side, it looks like it render all at once instead of gradually. Who knows :smile:

**Comment by TheLarkInn** [comment]
Awesome. Does this automatically get updated on HNPWA or do you have to PR that also? (Giddy to see the numbers they generate)

**Comment by WebReflection** [comment]
let me check numbers before even mentioning I've changed pretty much everything underneath :smile:

**Comment by WebReflection** [comment]
almost there, same result:
blob:chrome-extension://blipmdconlkpinefehnmjammfjpmpbjk/9f9d8d0e-1b51-4355-b59a-54e7d9e88eb4

**Comment by WebReflection** [comment]
I keep posting pointless links, sorry

**Comment by TheLarkInn** [comment]
For somereason I can never load those blobs.

**Comment by TheLarkInn** [comment]
Haha it's okay. So is it still a regression vs original? Or faster.

**Comment by WebReflection** [comment]
dunno yet, I'm deploying last layout which is this one:
```html
<!doctype html>
<html lang="en">
  <head>
    <title> ${model.title} </title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="theme-color" content="#ffffff">
    <style>${model.style}</style>
    <link rel="manifest" href="/manifest.json">
    <link rel="preload" href="/js/bundle.js" as="script">
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="/favicon-16x16.png" sizes="16x16">
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
  </head>
  <body>
    <header>${model.header}</header>
    <main>${model.main}</main>
    <script async src="/js/bundle.js"></script>
  </body>
</html>
```

Once I've done that, I might leave as it is and compare other numbers through WebpagePerf.

After that, I'll try to figure out new numbers and I'll eventually PR.

I just want to be sure there's not a big regression. It was a very good exercise anyway, and it pushed me to bring in asynchronous chunks on the client side out of the box.

It's a win-win, numbers a part.

**Comment by TheLarkInn** [comment]
Awesome!!!!! Ping me here or on social once you get it all.sorted out. Very excited to see the results.

**Comment by WebReflection** [comment]
Btw, the fact that everything loads on demand while surfing is awesome.

The SW also doesn't seem to affect at all initial paint/layout.

screenshot from 2017-06-28 20-27-39
